packages/element/src/create-interpolate-element.js

```diff
@@ -0,0 +1,235 @@
+ *                                expression.
+ * @param {string}	type					What type of regEx to return.
+ *
+ * @return {RegExp}  The generated regular expression.
+ */
+const getRegEx = ( tags, searchString, type ) => {
+	let pattern;
+	const [ openTag, closeTag, selfClosingTag ] = tags;
+	switch ( type ) {
+		case 'children':
+			pattern = escapeRegExp( openTag ) + '(.*)' + escapeRegExp( closeTag );
+			break;
+		case 'propValue':
+			pattern = escapeRegExp( searchString );
+			break;
+		default:
+			pattern = escapeRegExp( selfClosingTag );
+	}
+	return new RegExp( pattern );
+};
+
+/**
+ * Generates and returns the regular expression used for splitting a string
+ *
+ * @param {string} searchString  The search string serving as the base for the
+ *                               expression.
+ *
+ * @return {RegExp}  The generated regular expression
+ */
+const getSplitRegEx = ( searchString ) => new RegExp( '(' + escapeRegExp( searchString ) + ')' );
+
+/**
+ * Generates a String.prototype.match value for the incoming arguments.
+ *
+ * @param {string} interpolatedString  The string the match is performed on.
+ * @param {string} searchString				 The string used as the base for the
+ *                                     search
+ * @param {Object} conversionConfig    The configuration object for the
+ *                                     interpolation being performed.
+ *
+ * @return {Array|null}	An array if there is a match or null if not.
+ */
+const getMatchFromString = (
+	interpolatedString,
+	searchString,
+	conversionConfig
+) => {
+	const tags = getTagsFromString( searchString );
+	// first try children reg ex.  If there is a match, then return.
+	const match = interpolatedString.match(
+		getRegEx( tags, searchString, 'children' )
+	);
+	if ( match !== null ) {
+		conversionConfig.hasChildren = true;
+		return match;
+	}
+	// no children, try matching on another pattern and return.
+	const regEx = getHasPropValue( conversionConfig ) ?
+		getRegEx( tags, searchString, 'propValue' ) :
+		getRegEx( tags, searchString );
+	return interpolatedString.match( regEx );
+};
+
+// index for keys
+// This is external to `recursiveCreateElement` and reset in
+// `createInterpolateElement` because of the recursion.
+let keyIndex = -1;
+
+/**
+ * Used to recursively create elements from the interpolation string using the
+ * conversion map.
+ *
+ * @param {string}  potentialElement  The interpolation string (or fragment)
+ *                                    being processed.
+ * @param {Array[]} conversionMap     The interpolation map used for converting
+ *                                    the string to a react element.
+ *
+ * @return {Element|string|Array}  A react element, string or array.
+ */
+const recursiveCreateElement = ( potentialElement, conversionMap ) => {
+	/**
+	 * If the conversion map is not a valid array or empty then just return the
+	 * element.
+	 */
+	if ( ! Array.isArray( conversionMap ) || ! conversionMap.length ) {
```

Comment:
See my other comment about validating `conversionMap` in `createInterpolateElement` beforehand. With that in place, this check could be reduced to just the length condition.

Reply:
My reply to your validation comment applies here.  I have a feeling I'm not comprehending what you had in mind though.

Reply:
In the other comment thread, I think your argument that it would add an unnecessary loop over `conversionMap` is valid. In this case here, though, we could check once in `createInterpolateElement` that `Array.isArray( conversionMap )` and avoid repeating that test in each `recursiveCreateElement`.

Reply:
`conversionMap` is mutated, so there's still the necessity for verifying that it is valid before continuing to execute.